### PR TITLE
Incorporate yaml header generator into CMake

### PR DESCRIPTION
**Pull request by AzulRadio**
NOTE: Don't merge until yaml merges into main.

test: 
```
cd build
cmake ..
make -j
```
You should be finding generated headers in `vehicles/config`

### CODE REVIEW

File: vehicles/config/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+#                                                                        #
+#  You should have received a copy of the GNU General Public License     #
+#  along with this program. If not, see <http://www.gnu.org/licenses/>.  #
+#                                                                        #
+# ---------------------------------------------------------------------- #
+
+# include all yaml files as dependence
+file(GLOB_RECURSE YAML_FILES ${CMAKE_CURRENT_SOURCE_DIR}/*.yaml)
+
+# real all yaml files from yaml_list
```
Comment (0-EricZhou-0) [inline vehicles/config/CMakeLists.txt:24]:
```suggestion
# read all yaml files from yaml_list
```

**Comment by 0-EricZhou-0** [comment]
Now the configs are generated using all of the contents in the yaml files, could we change it to contain many different kinds of presets, and are tagged by an identifier, and that particular preset is used in the generation. For example, gimbal.yaml could be changed like this
```
Identifier1:
    pitch:
        <...>
    yaw:
        <...>

Identifier2:
    pitch:
        <...>
    yaw:
        <...>

...
```
Then the yaml to config generator could use the identifier either defined in the target, or in the files that are compiled to generate the configs. 

In this case, we can have several configs coexisting at the same time, and we select which one to use in compiling. Also this would prevent the files are updated too frequent on the branches.

**Comment by AzulRadio** [comment]
@0-EricZhou-0 TLDR: No. 
The idea is to **write different presets in different yaml files** in `vehicles/config/config_files`. Write different `yaml_list` files to designate which of the yaml files should be selected for header generation. 

Example: 
yaml_list: 
```
config_files/chassis.yaml
config_files/gimbal.yaml
config_files/shooter.yaml
```
Then the 3 selected yaml files will generate corresponding headers.

P.S. `yaml_list` is the param of the generation script. Usage: `python ./generate_all.py yaml_list`

**Comment by 0-EricZhou-0** [comment]
Now the configs are generated using all of the contents in the yaml files, could we change it to contain many different kinds of presets, and are tagged by an identifier, and that particular preset is used in the generation. For example, gimbal.yaml could be changed like this
```
Identifier1:
    pitch:
        <...>
    yaw:
        <...>

Identifier2:
    pitch:
        <...>
    yaw:
        <...>

...
```
Then the yaml to config generator could use the identifier either defined in the target, or in the files that are compiled to generate the configs. 

In this case, we can have several configs coexisting at the same time, and we select which one to use in compiling. Also this would prevent the files are updated too frequent on the branches.

> @0-EricZhou-0 TLDR: No.
> 
> The idea is to **write different presets in different yaml files** in `vehicles/config/config_files` and **select** which of the yaml files should be used for header generation **with different `yaml_list` files**.
> 
> Example: yaml_list:
> 
> ```
> config_files/chassis.yaml
> config_files/gimbal.yaml
> config_files/shooter.yaml
> ```
> 
> Then the 3 selected yaml files will generate corresponding headers.
> 
> P.S. `yaml_list` is the param of the generation script. Usage: `python ./generate_all.py yaml_list`

I see the yaml_list is hard-coded in the cmake files, how are we going to auto-change that according to the target we make?

**Comment by AzulRadio** [comment]
@0-EricZhou-0 I don't have a good idea so far. The best I can think of is to change yaml_list. Since each robot's mechanical configuration doesn't change often I think yaml_list can be hardcoded and different across different robots.

**Comment by AzulRadio** [comment]
Because we unavoidably will have something different in the code for different robots with different configurations (at least I can't think of a way to resolve it), I think this minor difference is reasonable and fine.

**Comment by 0-EricZhou-0** [comment]
@AzulRadio I have two ideas now:

1. Generate all the configs available in the `config_files/` directory and upload these generated configs to Github altogether, and user decide which one to include locally. The advantage of this method is that it can be achieved by the infrastructure you have right now. The drawback is that it would make the directory messy (but I think we do not expect user to read through that directory so it does not matter). 

2. Create global variables in cmake using `set`, and the python script would take in arguments from the cmake commands to make file generation more automatic. The advantage of this is that the directory is cleaner, but it has the drawback that these config files are changed from time to time, and from branch to branch. The most important drawback is that on Github we have build tests, and that will report error since none or only a subset of config files are uploaded to Github. 

I personally prefer the first method.

**Comment by AzulRadio** [comment]
@0-EricZhou-0 
Let's go with your first idea and put many different YAML files in the `config_files/`. Still, I think switching from different headers with different names can be chaotic for the CPP part. The latest commit [34c4d39](https://github.com/illini-robomaster/iRM_Embedded_2022/pull/39/commits/34c4d39a216dc5fdd267081ef42a6afbb06eefbf) now allows users to generate the same header file from different YAML files with `header_name` parameters defined in YAML files. But still, I don't want these generated headers to be added to the repo because that will be nasty.

**Comment by 0-EricZhou-0** [comment]
@AzulRadio Being able to change the header name is great, but how would you handle the Github build tests? And also, if we keep the config files like this, we would still need to automate the generation logic through cmake facilities. 

Add onto that, we already have tons of generated files in the repo, and I do not think it would be a problem if these generated files are also uploaded. They resides in an area that no human would try to read anyway.

**Comment by AzulRadio** [comment]
> how would you handle the Github build tests? 
> we would still need to automate the generation logic somehow through cmake facilities

I am not entirely sure what you are talking about. A big `cmake ..` `make -j` in `build/` can generate the headers and compile the whole project now. You are probably suggesting something already implemented.

> generated files in the repo

For different robots, we either use 
1. different headers with different names.
2. different headers from different Yaml with the same name.

We are going with option 2 here: the problem with option 1 is that cpp files needs to be changed on different robots and might cause problems during merging; for option 2, we will have different content under the same name in different robots, if we added these headers to the repo, we still have a big problem when we try to merge. Conclusively, the current solution is optimal.

Feel free to ask if there's something unclear.

**Comment by 0-EricZhou-0** [comment]
@AzulRadio I see what you are doing now. 

I still suggest that instead of using contents in the `yaml_list`, the python script should generate the headers using all the yaml file it can find in the `config_files/` folder. And yes, I agree that we should not upload these generated files, it's my bad that I forgot in the build test, `make -j` can generate these config files. 

If we choose to generate all the header files, naming the generated files according to the yaml file names would be an advantage. Because the yaml files are under the same dir, their names cannot duplicate; thus prevent overwriting.

**Comment by AzulRadio** [comment]
> generate the headers using all the yaml files it can find in the config_files/ folder.

No. 
Example: Say that we have `Infantry_1_gimbal.yaml` and `Infantry_2_gimbal.yaml` which both generate `GIMBAL_CONFIG.h` . The generation order is alphabetic, so the `GIMBAL_CONFIG.h` generated by `Infantry_1_gimbal.yaml` will be overwritten. 

Generating both `Infantry_1_gimbal.h` and `Infantry_2_gimbal.h` won't work either because that requires us to change things in cpp codes.

> Although the naming would be complicated and potentially ugly.

I've defined that if the header_file parameter is not defined or left blank, the generated header file will use the name of the YAML file. 

Feel free to ask if there's anything still unclear.

**Comment by 0-EricZhou-0** [comment]
@AzulRadio 
> Generating both `Infantry_1_gimbal.h` and `Infantry_2_gimbal.h` won't work either because that requires us to change things in cpp codes. (-> Problems in merging)

I think the solution to this is to give the robot parts universal and proper names. Each part will have its own name according to the year it is built, what kind of robot the part is placed on, etc. We can place labels with these universal names on the physical devices to make sure the user know which config files to change. For example `Infantry2022_A_gimbal.yaml` and `Infantry2022_B_gimbal.yaml` generate `INFANTRY2022_A_GIMBAL.h` and `INFANTRY2022_B_GIMBAL.h` respectively. Because we give all the parts identifiers, users can know which gimbal they are tuning, and write to corresponding yaml file(s). The exact naming scheme could be further discussed. 

I am trying to automate the entire process here, the configuration now would need us to change the `yaml_list` manually before generating, which I think is a little bit redundant and could be easily forgotten. I think we can make the whole process automated just by enforcing strict naming.

**Comment by AzulRadio** [comment]
@0-EricZhou-0 I thought about making a cmake function to do the auto-generation for every different robot and I also have a plan simpler than yours without enforcing strict naming. But the true problem isn't that -- the true problem is we have codes in shared/ that rely on YAMLs.

**Comment by AzulRadio** [comment]
@0-EricZhou-0 Again the `yaml_list` need no editing unless you are switching back and forth between several different robots at the same time or merging with other robots.

**Comment by 0-EricZhou-0** [comment]
@AzulRadio 
> We are planning to move the gimbal.cc etc. to dependent on YAMLs

Yes, and that is the reason we bring yaml config files in. Can't the files in `shared/` access the headers in another directory?

> Again the yaml_list need no editing unless you are switching back and forth between several different robots at the same time or merging with other robots.

I think there would be a time in the future that we might do that. The whole point that I want to automate the whole process is that the user might forget to change `yaml_list` when debugging or flashing, because that is something extra apart from `make flash_XXX` or `make debug_XXX`.

> I also have a plan simpler than yours without enforcing strict naming. 

I would be interested in that. What's your plan?

**Comment by AzulRadio** [comment]
> user might forget to change

good point.

The plan is to make the script a cmake function and make separate config yamls to generate headers in every robot's folders independently but not put all configs together (like examples/ but in vehicles/). A problem about this plan is I've not yet figured out where to put the configs for those in share/

**Review by 0-EricZhou-0 — APPROVED** [review]

**Pull request merged**